modules/graphdb/templates/08_cluster_rejoin.sh.tpl

```diff
@@ -0,0 +1,141 @@
+
+  # Waits for all nodes to be available (Handles rolling upgrades)
+  for node in "$${NODES[@]}"; do
+    while ! check_gdb "$node.$${DNS_ZONE_NAME}"; do
+      echo "Waiting for GDB $node.$${DNS_ZONE_NAME} to start"
+      sleep 5
+    done
+  done
+
+  # Iterates all nodes and looks for a Leader node to extract its address.
+  while [ -z "$LEADER_NODE" ]; do
+    for node in "$${NODES[@]}"; do
+      endpoint="http://$node.$${DNS_ZONE_NAME}:7200/rest/cluster/group/status"
+      echo "Checking leader status for $node.$${DNS_ZONE_NAME}"
+
+      # Gets the address of the node if nodeState is LEADER, grpc port is returned therefor we replace port 7300 to 7200
+      LEADER_ADDRESS=$(curl -s "$endpoint" -u "admin:$${GRAPHDB_ADMIN_PASSWORD}" | jq -r '.[] | select(.nodeState == "LEADER") | .address' | sed 's/7300/7200/')
+      if [ -n "$${LEADER_ADDRESS}" ]; then
+        LEADER_NODE=$LEADER_ADDRESS
+        echo "Found leader address $LEADER_ADDRESS"
+        break 2 # Exit both loops
+      else
+        echo "No leader found at $node"
+      fi
+    done
+
+    echo "No leader found on any node. Retrying..."
+    sleep 5
+  done
+
+  echo "Attempting to rejoin the cluster"
+  # Step 1: Remove the node from the cluster.
+  echo "Attempting to delete $${CURRENT_NODE_NAME}.$${DNS_ZONE_NAME}:7300 from the cluster"
+
+  DELETE_NODE=$(
+    curl -X DELETE -s \
+      -o "/dev/null" \
+      -H 'Content-Type: application/json' \
+      -H 'Accept: application/json' \
+      -w "%%{http_code}" \
+      -u "admin:$${GRAPHDB_ADMIN_PASSWORD}" \
+      -d "{\"nodes\": [\"$${CURRENT_NODE_NAME}.$${DNS_ZONE_NAME}:7300\"]}" \
+      "http://$${LEADER_NODE}/rest/cluster/config/node"
+  )
+
+  if [[ "$DELETE_NODE" == 200 ]]; then
+    echo "$${CURRENT_NODE_NAME}.$${DNS_ZONE_NAME} was successfully deleted from the cluster"
+  elif [[ "$DELETE_NODE" == 400 ]]; then
+    echo "Node was not found in the cluster config"
+  else
+    echo "Something went wrong, check the logs!"
+  fi
+
+  # Step 2: Add the current node to the cluster with the same address.
+  echo "Attempting to add $${CURRENT_NODE_NAME}.$${DNS_ZONE_NAME}:7300 to the cluster"
+  # This operation might take a while depending on the size of the repositories.
+  CURL_MAX_REQUEST_TIME=21600 # 6 hours
+
+  ADD_NODE=$(
+    curl -X POST -s \
+      -m $CURL_MAX_REQUEST_TIME \
+      -o "/dev/null" \
+      -H 'Content-Type: application/json' \
+      -H 'Accept: application/json' \
+      -w "%%{http_code}" \
+      -u "admin:$${GRAPHDB_ADMIN_PASSWORD}" \
+      -d"{\"nodes\": [\"$${CURRENT_NODE_NAME}.$${DNS_ZONE_NAME}:7300\"]}" \
+      "http://$${LEADER_NODE}/rest/cluster/config/node"
+  )
+
+  if [[ "$ADD_NODE" == 200 ]]; then
+    echo "$node.$${DNS_ZONE_NAME} was successfully added to the cluster."
+  else
+    echo "Node $${CURRENT_NODE_NAME}.$${DNS_ZONE_NAME} failed to rejoin the cluster, check the logs!"
+  fi
+}
+
+# Check if the Raft directory exists
+if [ ! -d "$RAFT_DIR" ]; then
```

Comment:
can't we replace this wait with a check with the other nodes and see if there's a cluster - if there is and the current node is in the node list but doesn't have a raft directory we rejoin it?

Reply:
We could, but would be very flaky during the initial cluster setup.